HDDS-3612. Document details of bucket mount design

adoroszlai opened the pull request:
## What changes were proposed in this pull request?

Add proposed implementation details to Ozone S3 volume mapping design doc.  The primary goal is to be able to discuss it before going forward with big code changes.

https://issues.apache.org/jira/browse/HDDS-3612

## How was this patch tested?

Checked preview on GitHub.

arp7 commented on hadoop-hdds/docs/content/design/ozone-volume-management.md:
```diff
@@ -4,7 +4,7 @@ summary: A simplified version of mapping between S3 buckets and Ozone volume/buc
 jira: HDDS-3331
 status: accepted
-author: Marton Elek, Arpit Agarwall, Sunjay Radia
+author: Marton Elek, Arpit Agarwal, Sanjay Radia
```
You can remove my name, I really cannot claim any authorship credit for this idea. 🙂

elek replied:
@arp7 You wrote additional section to this document about the volume ownership problems.

arp7 commented on hadoop-hdds/docs/content/design/ozone-volume-management.md:
```diff
@@ -106,19 +106,27 @@ This is an easy an fast method, but with this approach not all the volumes are a
+
+ * Let bucket mount operation create a link bucket.  Links are like regular buckets, stored in DB the same way, but with two new, optional pieces of information: source volume and bucket.
+ * Existing bucket operations (info, delete, ACL) work on the link object in the same way as they do on regular buckets.  No new link-specific RPC is required.
+ * Links are followed for key operations (list, get, put, etc.).  Checks for existence of the source bucket, as well as ACL, are performed at this time (similar to symlinks).  This avoids the need for reverse checks for each bucket delete or ACL change.
```
Yeah this should work a lot like symlinks, so we shouldn't perform reverse checks on changes to the target bucket.

arp7 commented on hadoop-hdds/docs/content/design/ozone-volume-management.md:
```diff
@@ -106,19 +106,27 @@ This is an easy an fast method, but with this approach not all the volumes are a
+ * Let bucket mount operation create a link bucket.  Links are like regular buckets, stored in DB the same way, but with two new, optional pieces of information: source volume and bucket.
+ * Existing bucket operations (info, delete, ACL) work on the link object in the same way as they do on regular buckets.  No new link-specific RPC is required.
+ * Links are followed for key operations (list, get, put, etc.).  Checks for existence of the source bucket, as well as ACL, are performed at this time (similar to symlinks).  This avoids the need for reverse checks for each bucket delete or ACL change.
+ * The same permission is required on both the link and the source bucket to be able to perform the operation via the link.  This allows finer-grained access control.
```
We should probably try to match the behavior of Unix symlinks wrt permissions.

arp7 commented on hadoop-hdds/docs/content/design/ozone-volume-management.md:
```diff
@@ -106,19 +106,27 @@ This is an easy an fast method, but with this approach not all the volumes are a
+
+#### Implementation details
+
+ * Let bucket mount operation create a link bucket.  Links are like regular buckets, stored in DB the same way, but with two new, optional pieces of information: source volume and bucket.
```
> Let bucket mount operation create a link bucket
Didn't understand this sentence. Does it mean that when you try to mount a bucket in a new volume it silently creates a link under the covers? Is the link reused next time we try to mount again?

Also how do we handle name collisions? Can the user choose any name for the link/mount point?

adoroszlai replied:
> > Let bucket mount operation create a link bucket
> 
> Didn't understand this sentence. Does it mean that when you try to mount a bucket in a new volume it silently creates a link under the covers?

Mount and link are the same.  (Even the high-level parts of the doc refer to it as "link" in some places and "mount" in others.  Maybe we should not mix terminology.  Since proposed behavior is closer to symlinks, I suggest using "link" instead of "mount".)

> Is the link reused next time we try to mount again?

The mount is "active" until the link is deleted via regular DeleteBucket request.  Until then, another attempt to mount (or create a bucket with same name) will result in a "bucket already exists" error.

> Also how do we handle name collisions? Can the user choose any name for the link/mount point?

Name collision is handled the same as if trying to create a regular bucket.

Basically, the extra link info does not change behavior until client starts working with keys in the "bucket".  Then it is redirected.

adoroszlai commented:
Thanks @arp7 for reviewing this doc.  I have updated it with some clarifications.  Also changed to use "symbolic link" terminology instead of "mount" for consistency.  This changes the text of the original proposal very slightly, but keeps its spirit/intentions, I think.  Eg. propose to call the command `ozone sh bucket link` instead of `ozone sh mount`.

adoroszlai commented:
Thanks @arp7 and @elek for the review.
